tools/generate_tables_sql_and_testNames.php

```diff
@@ -1,118 +1,234 @@
-                    $bits[0]=enumizeOptions($bits[3], isset($tablename) ? $tablename : null, $bits[1]);
-                } else if($bits[0]=="selectmultiple"){
-                    $bits[0]="varchar(255)";
-                } else if($bits[0]=="textarea"){
-                    $bits[0]="text";
-                } else if($bits[0]=="text"){
-                    $bits[0]="varchar(255)";
-                } else if($bits[0]=="checkbox") {
-                    $bits[0]="varchar(255)";
-                } else if ($bits[0]=="static") {
-                    $bits[0]="varchar(255)";
+                if ($bits[0] == "select") {
+                    $bits[0] = enumizeOptions($bits[3]);
+                } else if ($bits[0] == "selectmultiple") {
+                    $bits[0] = "varchar(255)";
+                } else if ($bits[0] == "textarea") {
+                    $bits[0] = "text";
+                } else if ($bits[0] == "text") {
+                    $bits[0] = "varchar(255)";
+                } else if ($bits[0] == "checkbox") {
+                    $bits[0] = "varchar(255)";
+                } else if ($bits[0] == "static") {
+                    $bits[0] = "varchar(255)";
+                } else {
+                    // Sanitize $bits to mitigate SQL injection attacks.
+                    $bits[0] = str_replace('`', '', $db->escape($bits[0]));
                 }
-
-                $bits[2]=htmlspecialchars($bits[2]);
-                $output.="`$bits[1]` $bits[0] default NULL,\n";
+                $bits[2]     = htmlspecialchars($bits[2]);
+                $output     .= "`$sql_query_table_name` $bits[0] default NULL,\n";
+                $column_name = str_replace('`', '', $db->escape($bits[1]));
+                $sql_query_columns[] = "`$column_name` $bits[0] default NULL,\n";
+            }
         }
-
+        $output .= "PRIMARY KEY  (`CommentID`)\n) ENGINE=InnoDB"
+            ." DEFAULT CHARSET=utf8;\n";
+        $fp      = fopen($file_name, "w");
+        fwrite($fp, $output);
+        // Check if Table Name exists
+        if (empty($sql_query_table_name)) {
+            $error_message = "Table name does not exist";
+        } else {
+            fwrite(
+                $fp,
+                "REPLACE INTO test_names (Test_name, Full_name, Sub_group) "
+                ."VALUES ('$sql_query_table_name', '$title', 1);\n"
+            );
+            // Writes a record of the changes to a file but is not sourced directly.
+            foreach ($pages as $page_number => $page) {
+                fwrite(
+                    $fp,
+                    "INSERT INTO instrument_subtests "
+                    ."(Test_name, Subtest_name, Description)"
+                    ." VALUES ('$sql_query_table_name', '"
+                    .$sql_query_table_name."_page".($page_number+1)
+                    ."', '$page');\n"
+                );
+            }
+        }
+        // Writes the error message to the sql file.
```

Comment:
but why ??

error message should just be printed and file discarded. I dont think the sql file should be used as a log

Reply:
I think the reason was logging potential errors from users attempting to execute instrument files if my memory serves me correctly. The server will output an Loris exception to the user and someone could potentially look in the apache logs but I thought it would be easier to save the query that resulted in an error. It can be discarded if you think it should.

Reply:
I def dont like it no